benchmark/benchmark.cpp

```diff
@@ -0,0 +1,377 @@
+//
+// Copyright (c) 2015-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#include<nudb/test/test_store.hpp>
+
+#if WITH_ROCKSDB
+#include "rocksdb/db.h"
+
+char const* rocksdb_build_git_sha="Benchmark Dummy Sha";
+char const* rocksdb_build_compile_date="Benchmark Dummy Compile Date";
+#endif
+
+#include <boost/program_options.hpp>
+#include <boost/system/system_error.hpp>
+
+#include <array>
+#include <chrono>
+#include <cmath>
+#include <iomanip>
+#include <map>
+#include <memory>
+#include <random>
+#include <set>
+#include <utility>
+
+struct Timer
+{
+    using clock = std::chrono::steady_clock;
+    using time_point = clock::time_point;
+    time_point start_;
+
+    Timer() : start_(clock::now())
+    {
+    }
+
+    auto
+    elapsed() const
+    {
+        return std::chrono::duration_cast<std::chrono::duration<double>>(
+            clock::now() - start_);
+    }
+};
+
+template <class Generator, class F>
+std::chrono::duration<double>
+time_block(std::uint64_t n, Generator&& g, F&& f)
+{
+    Timer timer;
+    while(n--)
+    {
+        f(g());
+    }
+    return timer.elapsed();
+}
+
+class gen_key_value
+{
+    nudb::test::test_store& ts_;
+    std::uint64_t cur_;
+
+public:
+    gen_key_value(nudb::test::test_store& ts,
+        std::uint64_t cur)
+        : ts_(ts),
+        cur_(cur)
+    {
+    }
+    auto
+    operator()()
+    {
+        return ts_[cur_++];
+    }
+};
+
+class rand_existing_key
+{
+    nudb::test::xor_shift_engine rng_;
+    std::uniform_int_distribution<std::uint64_t> dist_;
+    nudb::test::test_store& ts_;
+
+public:
+    rand_existing_key(nudb::test::test_store& ts,
+        std::uint64_t max_index,
+        std::uint64_t seed = 1337)
+        : ts_(ts),
+          dist_(0, max_index)
+    {
+        rng_.seed(seed);
+    }
+    auto
+    operator()()
+    {
+        return ts_[dist_(rng_)];
+    }
+};
+
+#if WITH_ROCKSDB
+std::map<std::string, std::chrono::duration<double>>
+do_timings_rocks(std::uint64_t num_inserts,
+    std::uint64_t num_fetches,
+    std::uint32_t key_size)
+{
+    std::map<std::string, std::chrono::duration<double>> result;
+    nudb::test::temp_dir td;
+
+    std::unique_ptr<rocksdb::DB> pdb = [path = td.path()]
```

Comment:
You can cherry pick this to get benchmark to compile with C++-11: https://github.com/seelabs/nudb/commit/906a05acf6bfc2a9cc9cff4a3aa4cb691d854e72